### PR TITLE
add dynamic padding substitution for nn.ConvTranspose2d

### PR DESCRIPTION
**Pull request by eladc-git**
## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/convtranspose_dynamic_padding.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+from model_compression_toolkit.core.pytorch.constants import OUTPUT_PADDING
+from model_compression_toolkit.core.common.graph.graph_matchers import NodeOperationMatcher
+from model_compression_toolkit.core import common
+from model_compression_toolkit.core.common import BaseNode, Graph
+from model_compression_toolkit.logger import Logger
+
+
+class ConvtransposeDynamicPadding(common.BaseSubstitution):
+    """
+    Replace output_padding of nn.convtranspose to align dynamic output_size input
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/convtranspose_dynamic_padding.py:27]: Explain the substitution better.

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/convtranspose_dynamic_padding.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    """
+
+    def __init__(self):
+        """
+        Matches: functional batch_norm
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/convtranspose_dynamic_padding.py:32]: ahmmm....

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/convtranspose_dynamic_padding.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                                                num_spatial_dims,
+                                                convtr.dilation)
+        return tuple(output_padding)
+
+
+    def substitute(self,
+                   graph: Graph,
+                   node: BaseNode) -> Graph:
+        """
+        Substitute functional.batch_norm and its inputs with BatchNorm2d.
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/convtranspose_dynamic_padding.py:63]: ahhmmm...

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/convtranspose_dynamic_padding.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        Args:
+            graph: Graph we apply the substitution on.
+            node: node that match the pattern in the substitution init.
+
+        Returns:
+            Graph after applying the substitution.
+        """
+        # Check that output is only contain a single tensor
+        if len(node.output_shape) > 1:
```
Comment (elad-c) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/convtranspose_dynamic_padding.py:72]: Conv always has 1 output. Can remove this check.

File: tests/pytorch_tests/model_tests/feature_models/convtranspose_dynamic_output_size_test.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+class ConvTranspose2dDynamicNet(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.downsample = nn.Conv2d(3, 16, 3, stride=2, padding=1)
+        self.upsample = nn.ConvTranspose2d(16, 3, 3, stride=2, padding=1)
+
+    def forward(self, x):
+        x = self.downsample(x)
+        x = self.upsample(x, output_size=[224, 224])  # <--- dynamic output_size
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/convtranspose_dynamic_output_size_test.py:33]: Add a test with `output_size=[225, 225]` so we cover all possible output sizes.

**Review by elad-c — CHANGES_REQUESTED** [review]

File: tests/pytorch_tests/model_tests/feature_models/convtranspose_dynamic_output_size_test.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import torch.nn as nn
+from tests.pytorch_tests.model_tests.base_pytorch_test import BasePytorchTest
+from model_compression_toolkit.core.pytorch.utils import to_torch_tensor
+
+"""
+This test checks dynamic output size for nn.ConvTranspose2d.
+"""
+
+
+class ConvTranspose2dDynamicNet(nn.Module):
```
Comment (haihabi) [inline tests/pytorch_tests/model_tests/feature_models/convtranspose_dynamic_output_size_test.py:25]: The test should be in the new format ask @ofirgo

**Review by elad-c — APPROVED** [review]
